### PR TITLE
feat(forge): prankStart + prankStop cheatcodes

### LINKED ISSUE #247: feat(prank): startPrank/stopPrank
would be dope to have a start and stop prank cheatcode that would allow impersonating an account over multiple txs so users won't have to repeat themselves before each

### PR DESCRIPTION
**Pull request by brockelmore**
Closes #247 

```solidity
interface Vm {
    function prank(address) external;
    function prankStart(address) external;
    function prankStop() external;
}

contract Prank {
    function bar(address expectedMsgSender) public {
        require(msg.sender == expectedMsgSender, "bad prank");
        InnerPrank inner = new InnerPrank();
        inner.bar(address(this));
    }
}

contract InnerPrank {
    function bar(address expectedMsgSender) public {
        require(msg.sender == expectedMsgSender, "bad prank");
    }
}

contract MyTest is DSTest {
    Vm vm = Vm(0x7109709ECfa91a80626fF3989D68f67F5b1DD12D);

    function testPrank() public {
        Prank prank = new Prank();
        address new_sender = address(1337);
        vm.prank(new_sender);
        prank.bar(new_sender);
        prank.bar(address(this));
    }

    function testPrankStart() public {
        Prank prank = new Prank();
        address new_sender = address(1337);
        vm.startPrank(new_sender);
        prank.bar(new_sender);
        prank.bar(new_sender);
        vm.stopPrank();
        prank.bar(address(this));
    }
}
```

### CODE REVIEW

File: evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs [hunk 1]
```diff
@@ -727,18 +763,30 @@ impl<'a, 'b, B: Backend, P: PrecompileSet> Handler for CheatcodeStackExecutor<'a
...
-            new_context.caller = caller;
-        }
-
         if code_address == *CHEATCODE_ADDRESS {
-            self.apply_cheatcode(input)
+            self.apply_cheatcode(input, context.caller)
         } else if code_address == *CONSOLE_ADDRESS {
             self.console_log(input)
         } else {
+            // apply some cheatcode relevant things
```
Comment (gakonst) [inline evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs:771]:
```suggestion
            // modify execution context depending on the cheatcode
```

File: evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs [hunk 2]
```diff
@@ -377,8 +377,44 @@ impl<'a, 'b, B: Backend, P: PrecompileSet> CheatcodeStackExecutor<'a, 'b, B, P>
             }
             HEVMCalls::Prank(inner) => {
                 let caller = inner.0;
+                if let Some((_orginal_pranker, _caller, depth)) = self.state().msg_sender {
+                    let start_prank_depth = if let Some(depth) = self.state().metadata().depth() {
+                        depth + 1
```
Comment (gakonst) [inline evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs:382]: pls add a comment on why we're checking callstack depth here

File: evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs [hunk 3]
```diff
@@ -377,8 +377,44 @@ impl<'a, 'b, B: Backend, P: PrecompileSet> CheatcodeStackExecutor<'a, 'b, B, P>
...
+                    } else {
+                        0
+                    };
+                    if start_prank_depth == depth {
+                        return evm_error("You have an active `startPrank` at this frame depth already. Use either `prank` or `startPrank`");
```
Comment (gakonst) [inline evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs:387]: I think?
```suggestion
                        return evm_error("You have an active `startPrank` at this frame depth already. Use either `prank` or `stopPrank`");
```

**Review by gakonst — APPROVED** [review]
LGTM - pending an extra test for the A->B{starPrank, doesn't end prank}->C edge case
